sangryeol/Week_7/시험장 나누기.java

```diff
@@ -0,0 +1,63 @@
+import java.util.*;
+
+public class Programmers_시험장나누기 {
+    public static void main(String[] args) {
+        Solution ret = new Solution();
+        System.out.println(ret.solution(3, new int[]{12, 30, 1, 8, 8, 6, 20, 7, 5, 10, 4, 1},
+                new int[][]{{-1, -1}, {-1, -1}, {-1, -1}, {-1, -1}, {8, 5}, {2, 10}, {3, 0}, {6, 1}, {11, -1}, {7, 4}, {-1, -1}, {-1, -1}}));
+    }
+}
+class Room {
+    int left, right;
+    int peopleCount;
+    Room(int left, int right, int peopleCount) {
+        this.left = left;
+        this.right = right;
+        this.peopleCount = peopleCount;
+    }
+}
+class Solution {
```

Comment:
[회고]

처음에 문제를 제대로 읽지 않은 상태로 간선을 끊는 그림만 보고.. 오프라인 쿼리로 union-find를 하는 문제인가.. 생각을 했다. 하지만 프로그래머스에서 주어지는 문제의 구조는 interactive한 형태이기 때문에 사실상 불가능하다.

문제를 제대로 읽어보니 **k - 1**개의 간선을 끊어 **k**개의 그룹을 만들어야 하고, 이 때 **가장 큰 그룹의 인원을 최소화 시키는 것**이 문제에서 요구하는 사항이다.

문제의 요구사항을 연습장에 끄적이다가 생각한 첫 번째 방법이다.

> dp를 쓰면 될려나..? dp[정점의 번호][0~2] 이런 식으로 상태관리를 하면 되지 않을까? 아 근데 k - 1개의 간선을 끊어서 k개의 그룹을 만들려면 k도 상태관리에 필요하구나.. 그럼 내가 아는 선에서는 dp를 쓰기에 메모리 초과가 발생하겠다..

위와 같은 생각을 하고 다음 방법을 생각해보니 가장 큰 그룹의 인원을 x라 할 때 생길 수 있는 **최소 그룹 수**가 단조 감소 함수 꼴인 것을 확인하였다. 그러면 파라메트릭 서치를 사용하는 문제구나..! 라고 생각을 했고 결정 함수를 어떻게 만들지 많은 고민을 하였다. 

> 한 그룹의 최대 인원 수가 x일 때, k 이하의 그룹이 만들어지는 확인할 수 있는 결정 함수

왜냐하면 어디서 간선을 끊을지 결정하는지에 따라 시간복잡도가 달라질 수 있으며 올바른 결정 함수가 아닐 수도 있기 때문이다. 그냥 모든 경우를 다 보면 2^(정점의 개수)인데 정점의 개수는 10000개이다..

고민을 계속 하다가 leaf node에서 위로 올라가는 식으로 탐색을 진행하게 되면 최소 그룹의 수를 구할 수 있지 않을까라는 생각을 하게 되었다. 즉 post-order로 트리를 탐색을 하면서 왼쪽 그룹에 속하는 인원 수와 오른쪽 그룹에 속하는 인원의 수로 case 분류를 해주면서 필요한 최소 그룹의 수를 계산해주었다.

```c
          root
       /        \
    left       right
```
1. `root + left + right <= x` : 새롭게 그룹이 생기지 않아도 되는 경우
2. `otherwise 1 and root + min(left, right) <= x` : max(left, right)의 간선을 cut하여 새롭게 그룹이 1개 더 생기는 경우
3. `otherwise 1, 2 and root + min(left, right) > x` : 새롭게 그룹을 2개 추가하는 경우

위와 같은 case 분류로 트리를 탐색하는 함수를 구현하였다.

20년도 상반기에 지원했던 카카오 인턴 문제에서도 학부 수준에서 배운 알고리즘에 대한 이해를 묻는 문제가 3,4,5번에 있었는데, 이 문제의 의도는 **파라메트릭 서치, 트리에서의 탐색**인 것 같다.. 스터디를 통해 굉장히 생각을 많이 하게 하는 좋은 문제를 풀었던 것 같다.